### PR TITLE
CSHARP-2151: Resync connection string spec tests for percent encoding in userinfo

### PR DESCRIPTION
**Pull request by ksadoff**
https://jira.mongodb.org/browse/CSHARP-2151
~~https://evergreen.mongodb.com/version/5cffdde19ccd4e5a3c8d7be6~~
~~https://evergreen.mongodb.com/version/5d02ad1d3066154a30d7f666~~
~~https://evergreen.mongodb.com/version/5d03b55f7742ae1530791b60~~
https://evergreen.mongodb.com/version/5d03f56a3e8e8660cc856f67

### CODE REVIEW

**Comment by ksadoff** [comment]
Ready for interim review

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 1]
```diff
@@ -63,7 +63,7 @@ public sealed class ConnectionString
         private readonly Dictionary<string, string> _authMechanismProperties;
         private readonly CompressorsOptions _compressorsOptions;
 
-        // these are all readonly, but since they are not assigned 
+        // these are all readonly, but since they are not assigned
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Ah, looks like the "trim trailing whitespace" setting has automatically fixed the extra whitespace on this line. I would probably revert this change as in this case, it adds some noise `git annotate`/`git blame` since it'd be more useful to see the commit when the line had a substantive change as opposed to the commit when whitespace was tweaked

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 2]
```diff
@@ -365,7 +365,7 @@ public IReadOnlyList<TagSet> ReadPreferenceTags
         {
             get { return _readPreferenceTags; }
         }
-        
+
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:368]: Ah, looks like the "trim trailing whitespace" setting has automatically fixed the extra whitespace on this line. While we tend to avoid making unnecessary changes to files, I don't mind the change here as there's `git blame` shows nothing useful on this line anyways, but I'll defer to @rstam on this.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 3]
```diff
@@ -1111,7 +1117,7 @@ private static TimeSpan ParseTimeSpan(string name, string value)
         }
 
         private List<string> GetHostsFromResponse(IDnsQueryResponse response)
-        {         
+        {
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:1125]: I think this whitespace change is fine as it fixes a previous style mistake and again, no useful information is lost (see comments on lines 66 and 368)

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 4]
```diff
@@ -1130,7 +1136,7 @@ private List<string> GetOptionsFromResponse(IDnsQueryResponse response)
         {
             var txtRecords = response.Answers
                 .TxtRecords().ToList();
-            
+
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:1144]: Ditto

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 5]
```diff
@@ -759,8 +759,14 @@ private void Parse()
                 @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
                 serversPattern + @"(/" + databasePattern + ")?/?" + optionsPattern + "$";
 
+            Match percentMatch = null;
+            if (_originalConnectionString.Contains("%"))
+            {
+                percentMatch = Regex.Match(_originalConnectionString, "(%[0-9A-F-a-f]{2})+");
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Glad to see the unit tests passing! I think there may a minor hiccup here though: consider the following string
`mongodb://user%20%:password@localhost`--I think this should be invalid?

One possible approach could be to check for the invalid combinations:
1. % followed a non hex character
2. % followed by a hex character and a non hex character
Off the cuff, I think this would loosely look like:
``(%[^0-9A-Fa-f])|(%[0-9A-Fa-f][^0-9A-Fa-f])``
You might also be able to use [string interpolation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) to make the regex vaguely easier to read by defining an intermediate value such as 
```
var hexDigit= "0-9A-Fa-f";
var pattern = $"(%[^{hexDigit}])|"  // % followed by a non-hexadecimal digit OR
    +$"(%[{hexDigit}][^{hexDigit}])"; // % followed by a hexadecimal digit and a non-hexadecimal digit
```

**Review by vincentkam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 6]
```diff
@@ -759,8 +759,17 @@ private void Parse()
                 @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
                 serversPattern + @"(/" + databasePattern + ")?/?" + optionsPattern + "$";
 
+            Match badPercentMatch = null;
+            var hexDigit= "0-9A-Fa-f";
+            var percentPattern = $"(%[^{hexDigit}])|"  // % followed by a non-hexadecimal digit OR
+                                 +$"(%[{hexDigit}][^{hexDigit}])"; // % followed by a hexadecimal digit and a non-hexadecimal digit
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Ah, sorry! We tend to use a four space-indent for multiline statements (and one space before and after operators like `+`)
The dev guide was unclear about this, and I've updated it accordingly

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 7]
```diff
@@ -759,8 +759,17 @@ private void Parse()
...
+            if (_originalConnectionString.Contains("%"))
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:762]: I think this line was previously necessary when we were exploring the other implementation option, but is it still necessary with the current approach?

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 8]
```diff
@@ -759,8 +759,17 @@ private void Parse()
                 @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
                 serversPattern + @"(/" + databasePattern + ")?/?" + optionsPattern + "$";
 
+            Match badPercentMatch = null;
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: It occurs to me that the spec is concerned with `%` only in the userinfo portion of the connection string. So my proposal is thus: perhaps it would be best to migrate this logic into a new helper function. Then we could create a ` ValidateUserNameAndPassword(...)` that's structured very similarly to `ExtractUserNameAndPassword(..)`.  `ValidateUserNameAndPassword(...)`  could call this new helper function, first checking the username and then the password, and throw an exception if the new helper function returns false. We could call this validation function prior to calling `ExtractUserNameAndPassword(...)`. Or something along those lines...  What do you think?

**Review by vincentkam — CHANGES_REQUESTED** [review]

**Comment by vincentkam** [comment]
Adding other reviewers as I think this is ready for review :)

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 8 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: That makes sense. I can go ahead and implement that if other reviewers think it would be helpful.

Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Would it make sense to call `ValidateUserNameAndPassword(...)`after we've extracted username and password? Or potentially from `ExtractUserNameAndPassword(...)`? Otherwise, it seems as though we'd either need to continue validating the whole connection string, or we'd be duplicating some of the work that gets done in `ExtractUserNameAndPassword(...)`.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 9]
```diff
@@ -759,8 +759,17 @@ private void Parse()
...
+            var hexDigit= "0-9A-Fa-f";
+            var percentPattern = $"(%[^{hexDigit}])|"  // % followed by a non-hexadecimal digit OR
+                                 +$"(%[{hexDigit}][^{hexDigit}])"; // % followed by a hexadecimal digit and a non-hexadecimal digit
+            if (_originalConnectionString.Contains("%"))
+            {
+                badPercentMatch = Regex.Match(_originalConnectionString, percentPattern);
+            }
+
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:772]: I would not mix the new validation (for valid % escape sequences) with the existing validation (for overall structure). Perhaps something like this at this location in the file:

    var invalidPercentPattern = @"%$|%.$|%[^0-9a-fA-F]|%[0-9a-fA-F][^0-9a-fA-F]";
    if (Regex.IsMatch(_originalConnectionString, invalidPercentPattern))
    {
        var message = string.Format("The connection string '{0}' contains an invalid '%' escape sequence.", _originalConnectionString);
        throw new MongoConfigurationException(message);
    }

I've added a few cases to the regular expression to cover percent signs near the end of the string that are not followed by the required two characters.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 10]
```diff
@@ -759,8 +759,17 @@ private void Parse()
...
             var match = Regex.Match(_originalConnectionString, pattern);
-            if (!match.Success)
+            if (!match.Success || (match.Success && badPercentMatch!=null && badPercentMatch.Success))
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: This if statement is complicated. In part because match.Success is in there twice.

If we validate % escape sequences separately as suggested above this line doesn't need to change.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 9 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:772]: p.s. It's possible that an approach not using regular expressions might be simpler but maybe not.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 7 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:762]: Not necessary, but a possible performance micro-optimization if connection strings hardly ever contain % escape sequences because this check would be faster than running an expensive regular expression check.

File: tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml [hunk 11]
```diff
@@ -191,3 +191,12 @@ tests:
...
         options: ~
+    -
+        description: "Username with password containing an unescaped percent sign"
+        uri: "mongodb://alice%foo:bar@127.0.0.1"
+        valid: false
+        warning: ~
+        hosts: ~
+        auth: ~
+        options: ~        
+
```
Comment (rstam) [inline tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml:202]: Possibly not in the scope of this ticket, but Visual Studio is not happy with the `connection-string\TestRunner` class. It won't let me select and debug individual tests unless I make the following changes to use the new `JsonDrivenTestCase` class which is Visual Studio/Xunit friendly:

Replace:

    public void RunTestDefinition(BsonDocument definition)

With:

    public void RunTestDefinition(JsonDrivenTestCase testCase)
    {
        var definition = testCase.Test;

And replace the entire `TestCaseFactory` class with:

    private class TestCaseFactory : JsonDrivenTestCaseFactory
    {
        private static readonly string[] __ignoredTestNames = new string[]
        {
            "invalid-uris.json:Missing delimiting slash between hosts and options"
        };

        protected override string PathPrefix => "MongoDB.Driver.Core.Tests.Specifications.connection_string.tests.";

        protected override IEnumerable<JsonDrivenTestCase> CreateTestCases(BsonDocument document)
        {
            return base.CreateTestCases(document)
                .Where(test => !__ignoredTestNames.Any(ignoredName => test.Name.EndsWith(ignoredName)));
        }
    }

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 8 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Doesn't seem like there is any advantage to only validating part of the connection string.

We support % escape sequences everywhere (not just username/password), so we might as well validate the whole connection string.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 8 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: In that case, I think it makes sense to leave the validation in the `Parse(...)` function.
The spec ticket (https://jira.mongodb.org/browse/SPEC-835) specifically mentions the password, but it seems as though "userinfo" refers to the whole connection string, so I think checking the whole URI for this ticket should be good.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 9 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:772]: @vincentkam and I tried to validate the connection string using the library function `Uri.IsWellFormedUriString(String, UriKind)` but that doesn't work with pre-existing test cases, so I think regular expressions may be our best bet.

File: tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml (hunk 11 above)
Comment (ksadoff) [reply tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml:202]: I've gone ahead and made this change, but we can revert it and leave it for another ticket if you think that's more appropriate.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 9 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:772]: That gave me an idea. The following helper method probably performs much faster than regular expressions:

    public static bool AllPercentEncodingsAreValid(string value)
    {
        for (var i = 0; i < value.Length; i++)
        {
            if (value[i] == '%' && !Uri.IsHexEncoding(value, i))
            {
                return false;
            }
        }

        return true;
    }

Or just to tease @vincentkam here's a slower functional version:

    public static bool AllPercentEncodingsAreValid2(string value)
        => Enumerable.Range(0, value.Length).All(i => value[i] != '%' || Uri.IsHexEncoding(value, i));

File: tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml (hunk 11 above)
Comment (rstam) [reply tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml:202]: I'm OK with it here as part of this ticket as long as new one else objects.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 9 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:772]: LOL at the functional version tease 😸 The docs do state that `All()` returns early so it should only be a tiny tiny bit slower due to the overhead of LINQ.

Also doable would be...
`s.Select((_, index) => index).All(index => s[index] != '%' || Uri.IsHexEncoding(s, index))` 😆

File: tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml (hunk 11 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml:202]: SGTM!

Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Specifications/connection-string/tests/invalid-uris.yml:202]: We were also seeing issues on Rider fwiw

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 7 above)
Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:762]: Sgtm! Resolving !

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 9 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs:772]: After offline discussion with @vincentkam, I decided to use an #ifdef for `AllPercentEncodingsAreValid(...)`, since the function [`Uri.IsHexEncoding(...)`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.ishexencoding?view=netframework-4.8#System_Uri_IsHexEncoding_System_String_System_Int32_) is not available in .NETStandard 1.5, which is what we're targeting. I kept the regex way as an alternative. We could also choose to vendor in the library function if that would make more sense.
And I decided to go with Robert's implementation of the helper function (sorry Vincent 😄 )

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 12]
```diff
@@ -747,6 +747,23 @@ private string GetHostNameForDns()
             return host;
         }
 
+#if !NETSTANDARD1_5
+        private static bool AllPercentEncodingsAreValid(string value)
+        {
+            for (var i = 0; i < value.Length; i++)
+            {
+                //Uri.IsHexEncoding is not in .NETStandard 1.5, which was my default on Rider
+                //also appears to not compile if I use .NETFramework
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: We like to put one space after "//" before the start of the text.

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: That's unfortunate about .NET Standard 1.5.

Do you think it's worth having two different implementations? Or should we just use regular expressions on both platforms?

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 13]
```diff
@@ -759,6 +776,24 @@ private void Parse()
...
 
+            if (_originalConnectionString.Contains("%"))
+            {
+                bool invalidPercentEncoding = false;
+#if NETSTANDARD1_5
+                var invalidPercentPattern = @"%$|%.$|%[^0-9a-fA-F]|%[0-9a-fA-F][^0-9a-fA-F]";
+                invalidPercentEncoding = Regex.IsMatch(_originalConnectionString, invalidPercentPattern);
+
+#else
+                invalidPercentEncoding = AllPercentEncodingsAreValid(_originalConnectionString);
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Don't you need to negate the return value of `AllPercentEncodingsAreValid`?

This should have caused a test to fail.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 14]
```diff
@@ -747,6 +747,23 @@ private string GetHostNameForDns()
             return host;
         }
 
+#if !NETSTANDARD1_5
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: @rstam I know you prefer `if bool` over `if !bool` : in this case though, it seems the .NET Standard 1.5 is the outlier as newer versions of .NET Standard (and surprisingly .NET Framework 4.5.2) have the needed function, so I feel like the `!NETSTANDARD1_5` is mildly more future proof. 

The alternative to `#if` would be to vendor in the `Uri.IsHexEncoding` function, which is available from the CoreFX project

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 14 above)
Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Let's not overthink it.

Once we've learned that `IsHexEncoding` is not available in .NET Standard 1.5 let's just abandon that approach and use the regular expression.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 13 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: That would explain why Evergreen is so unhappy :) I should have been more careful, let me fix that.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 12 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: I'm fine with simplifying things and sticking with the regular expression approach.

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs [hunk 15]
```diff
@@ -747,6 +747,20 @@ private string GetHostNameForDns()
...
+            bool invalidPercentEncoding;
+            var invalidPercentPattern = @"%$|%.$|%[^0-9a-fA-F]|%[0-9a-fA-F][^0-9a-fA-F]";
+            invalidPercentEncoding = Regex.IsMatch(value, invalidPercentPattern);
+
+            if (invalidPercentEncoding)
+            {
+                return false;
+            }
+
+            return true;
```
Comment (rstam) [inline src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Lines 756-761 seem odd. Why not just do this on line 754:

    return !Regex.IsMatch(value, invalidPercentPattern)

You might even consider renaming the helper method HasInvalidPercentEncodings or eliminating the helper method altogether.

    var invalidPercentPattern = @"%$|%.$|%[^0-9a-fA-F]|%[0-9a-fA-F][^0-9a-fA-F]";
    if (Regex.IsMatch(value, invalidPercentPattern))
    {
        // throw
    }

which is basically what we had originally right?

Comment (rstam) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: Sorry for the wild goose chase with IsHexEncoding... :)

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs (hunk 15 above)
Comment (ksadoff) [reply src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs]: I'm fine with moving the logic back into Parse(). It brings us back to where we were before, but I definitely learned stuff along the way, so no harm done!

**Review by rstam — APPROVED** [review]
LGTM

**Review by vincentkam — APPROVED** [review]
LGTM! 😺

**Comment by ksadoff** [comment]
Rebased, merged, and pushed.
